src/main/java/io/reactivex/internal/operators/observable/ObservableMergeWithMaybe.java

```diff
@@ -0,0 +1,262 @@
+
+    final MaybeSource<? extends T> other;
+
+    public ObservableMergeWithMaybe(Observable<T> source, MaybeSource<? extends T> other) {
+        super(source);
+        this.other = other;
+    }
+
+    @Override
+    protected void subscribeActual(Observer<? super T> observer) {
+        MergeWithObserver<T> parent = new MergeWithObserver<T>(observer);
+        observer.onSubscribe(parent);
+        source.subscribe(parent);
+        other.subscribe(parent.otherObserver);
+    }
+
+    static final class MergeWithObserver<T> extends AtomicInteger
+    implements Observer<T>, Disposable {
+
+        private static final long serialVersionUID = -4592979584110982903L;
+
+        final Observer<? super T> actual;
+
+        final AtomicReference<Disposable> mainDisposable;
+
+        final OtherObserver<T> otherObserver;
+
+        final AtomicThrowable error;
+
+        volatile SimplePlainQueue<T> queue;
+
+        T singleItem;
+
+        volatile boolean disposed;
+
+        volatile boolean mainDone;
+
+        volatile int otherState;
+
+        MergeWithObserver(Observer<? super T> actual) {
+            this.actual = actual;
+            this.mainDisposable = new AtomicReference<Disposable>();
+            this.otherObserver = new OtherObserver<T>(this);
+            this.error = new AtomicThrowable();
+        }
+
+        @Override
+        public void onSubscribe(Disposable d) {
+            DisposableHelper.setOnce(mainDisposable, d);
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (compareAndSet(0, 1)) {
+                actual.onNext(t);
+                if (decrementAndGet() == 0) {
+                    return;
+                }
+            } else {
+                SimplePlainQueue<T> q = getOrCreateQueue();
+                q.offer(t);
+                if (getAndIncrement() != 0) {
+                    return;
+                }
+            }
+            drainLoop();
+        }
+
+        @Override
+        public void onError(Throwable ex) {
+            if (error.addThrowable(ex)) {
+                DisposableHelper.dispose(mainDisposable);
+                drain();
+            } else {
+                RxJavaPlugins.onError(ex);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            mainDone = true;
+            drain();
+        }
+
+        @Override
+        public boolean isDisposed() {
+            return DisposableHelper.isDisposed(mainDisposable.get());
+        }
+
+        @Override
+        public void dispose() {
+            disposed = true;
+            DisposableHelper.dispose(mainDisposable);
+            DisposableHelper.dispose(otherObserver);
+            if (getAndIncrement() == 0) {
+                queue = null;
+                singleItem = null;
+            }
+        }
+
+        void otherSuccess(T value) {
+            if (compareAndSet(0, 1)) {
+                actual.onNext(value);
+                otherState = 2;
```

Comment:
Could we go and use some constants here to improve readability? It's really hard to follow.